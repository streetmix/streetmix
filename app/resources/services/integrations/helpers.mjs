import models from '../../../db/models/index.js'

const { User, UserConnections, sequelize } = models

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
export async function findUser (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
export function addOrUpdateByProviderName (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
export async function syncAccountStatus (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

/**
 * When a user has successfully connected their Streetmix account with a third-
 * party identity provider, record that connected identity with the user's
 * Streetmix data in the UserConnection table.
 *
 * @todo Remove first attempt at storing identity information as a JSON blob
 *        on a User model
 * @param {Object} account - User's Streetmix account information
 * @param {Object} profile - User's connected identity information
 * @returns {Promise}
 */
export function addUserConnection (account, profile) {
  const identities = account.identities
  const identity = {
    provider: profile.provider,
    user_id: profile.id,
    access_token: profile.access_token || undefined, // Coil
    refresh_token: profile.refresh_token || undefined // Coil
  }

  // if provider exists, should update that item
  // if not, should add to the object list
  // TODO: Deprecated
  addOrUpdateByProviderName(identities, identity)

  return sequelize.transaction(async (t) => {
    // This is a previous attempt, where we saved connected identity
    // information to a JSON field in the User model. A newer method (using
    // the UserConnection table) is used below. Both are currently used to
    // test concurrent data validity. In time, the User.update() method should
    // be removed.
    await User.update(
      {
        identities
      },
      { where: { auth0Id: account.auth0Id }, returning: true, transaction: t }
    )

    await UserConnections.findOrCreate({
      where: {
        user_id: account.id,
        provider: profile.provider
      },
      transaction: t
    })
    return await UserConnections.update(
      {
        provider_user_id: profile.id,
        metadata: profile
      },
      {
        where: {
          user_id: account.id,
          provider: profile.provider
        },
        returning: true,
        transaction: t
      }
    )
  })
}
