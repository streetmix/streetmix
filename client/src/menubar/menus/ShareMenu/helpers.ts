import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
