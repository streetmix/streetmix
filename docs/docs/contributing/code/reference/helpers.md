---
sidebar_position: 2
---

# Helpers and utilities

:::caution Under construction

This page is a work in progress.

:::

We use the following utilities to solve common programming problems.

## Pseudo-random number generator

If you need a one-time use random number (which will never be stored), there's nothing wrong with JavaScript's built-in `Math.random()`. However, there are times where we need a **seed** for a pseudo-random number generator (PRNG). Seeds generate a random-_looking_ sequence of numbers, but they have the benefit of remaining consistent across renders. For example, this is useful on our sidewalk segment, where we want to generate a random set of pedestrians but we want those pedestrians to remain the same between page refreshes or React component updates.

We can generate seeds by calling `generateRandSeed()`, then by feeding those seeds into the `seedrandom` library:

```js title="random-example.js"
import seedrandom from "seedrandom";
import { generateRandSeed } from "./util/random";

function doThingsWithThePRNG() {
  const seed = generateRandSeed();
  const prng1 = seedrandom(seed);

  // Generate pseudo-random numbers. Each time rng() is called, it returns
  // the next pseudo-random number in the sequence.
  console.log(prng1()); // -> first pseudo-random number
  console.log(prng1()); // -> second pseudo-random number
  // .. etc

  // Create another generator with the same seed
  const prng2 = seedrandom(seed);
  console.log(prng2()); // -> same value as the first call to prng1()
}
```

When `seedrandom()` is called, it returns a function that will return a consistent sequence of numbers each time it is called, for the given `seed`. The sequence returned is local to that function.

:::note

Although the `seed` can be any type of value, like a string, `generateRandSeed()` returns only integer values so that when we store seeds in a data model we can check for type consistency.

:::

In the above example, we call `seedrandom()` twice, with the same `seed`, so that we can create the same sequence of pseudo-random numbers in two different scenarios. This example is contrived because the pseudo-random number generators are used in the same function, and in reality the `seedrandom()` call may happen in separate parts of the application.

**References**

- [Random Seeds, Coded Hints, and Quintillions](http://davidbau.com/archives/2010/01/30/random_seeds_coded_hints_and_quintillions.html) [David Bau]
- [seedrandom.js](https://github.com/davidbau/seedrandom) [GitHub]

## axios

The [Axios](https://github.com/axios/axios) library was introduced as a recommendation over native `fetch` as a way to manage requests to the Streetmix back-end API layer. Additionally, Axios Mock Adapter offers the possibility to mock all API requests for integration testing.

If you work on API-related features, consider refactoring the `fetch` calls to use the API in `.assets/scripts/util/api.js`.
