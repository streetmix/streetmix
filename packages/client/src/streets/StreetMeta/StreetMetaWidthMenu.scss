.street-width-select {
  width: 150px;
}
