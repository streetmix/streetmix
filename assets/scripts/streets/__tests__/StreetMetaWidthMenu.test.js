/* eslint-env jest */
import React from 'react'
import { screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { render } from '../../../../test/helpers/render'
import StreetMetaWidthMenu from '../StreetMetaWidthMenu'

describe('StreetMetaWidthMenu', () => {
  it('renders (metric units, default width selected)', () => {
    const { asFragment } = render(
      <StreetMetaWidthMenu
        street={{
          units: 0,
          width: 60,
          occupiedWidth: 10
        }}
        onChange={jest.fn()}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders (metric units, custom width selected)', () => {
    const { asFragment } = render(
      <StreetMetaWidthMenu
        street={{
          units: 0,
          width: 10,
          occupiedWidth: 10
        }}
        onChange={jest.fn()}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders (imperial units, default width selected)', () => {
    const { asFragment } = render(
      <StreetMetaWidthMenu
        street={{
          units: 1,
          width: 60,
          occupiedWidth: 10
        }}
        onChange={jest.fn()}
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('calls onChange handler when selection changed', async () => {
    const handleChange = jest.fn((value) => value)
    const user = userEvent.setup()

    render(
      <StreetMetaWidthMenu
        street={{
          units: 0,
          width: 40,
          occupiedWidth: 10
        }}
        editable={true}
        onChange={handleChange}
      />
    )

    await user.selectOptions(screen.getByRole('combobox'), '40')

    // Return value is a string from <option value=""> attribute
    expect(handleChange).toHaveReturnedWith('40')
  })
})
