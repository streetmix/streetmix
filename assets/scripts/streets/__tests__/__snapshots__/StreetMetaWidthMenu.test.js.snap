// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreetMetaWidthMenu renders (imperial units, default width selected) 1`] = `
<DocumentFragment>
  <select
    class="street-width-select"
  >
    <option
      disabled=""
    >
      Occupied width:
    </option>
    <option
      disabled=""
    >
      10′
    </option>
    <option
      disabled=""
    />
    <option
      disabled=""
    >
      Building-to-building width:
    </option>
    <option
      value="40"
    >
      40′
    </option>
    <option
      value="60"
    >
      60′
    </option>
    <option
      value="80"
    >
      80′
    </option>
    <option
      value="-1"
    >
      Different width…
    </option>
    <option
      disabled=""
    />
    <option
      disabled=""
      id="switch-to-imperial-units"
      value="-3"
    >
      Switch to imperial units (feet)
    </option>
    <option
      id="switch-to-metric-units"
      value="-2"
    >
      Switch to metric units
    </option>
  </select>
</DocumentFragment>
`;

exports[`StreetMetaWidthMenu renders (metric units, custom width selected) 1`] = `
<DocumentFragment>
  <select
    class="street-width-select"
  >
    <option
      disabled=""
    >
      Occupied width:
    </option>
    <option
      disabled=""
    >
      3 m
    </option>
    <option
      disabled=""
    />
    <option
      disabled=""
    >
      Building-to-building width:
    </option>
    <option
      value="40"
    >
      12 m
    </option>
    <option
      value="60"
    >
      18 m
    </option>
    <option
      value="80"
    >
      24 m
    </option>
    <option
      disabled=""
    />
    <option
      value="10"
    >
      3 m
    </option>
    <option
      value="-1"
    >
      Different width…
    </option>
    <option
      disabled=""
    />
    <option
      id="switch-to-imperial-units"
      value="-3"
    >
      Switch to imperial units (feet)
    </option>
    <option
      id="switch-to-metric-units"
      value="-2"
    >
      Switch to metric units
    </option>
  </select>
</DocumentFragment>
`;

exports[`StreetMetaWidthMenu renders (metric units, default width selected) 1`] = `
<DocumentFragment>
  <select
    class="street-width-select"
  >
    <option
      disabled=""
    >
      Occupied width:
    </option>
    <option
      disabled=""
    >
      3 m
    </option>
    <option
      disabled=""
    />
    <option
      disabled=""
    >
      Building-to-building width:
    </option>
    <option
      value="40"
    >
      12 m
    </option>
    <option
      value="60"
    >
      18 m
    </option>
    <option
      value="80"
    >
      24 m
    </option>
    <option
      value="-1"
    >
      Different width…
    </option>
    <option
      disabled=""
    />
    <option
      id="switch-to-imperial-units"
      value="-3"
    >
      Switch to imperial units (feet)
    </option>
    <option
      id="switch-to-metric-units"
      value="-2"
    >
      Switch to metric units
    </option>
  </select>
</DocumentFragment>
`;
